sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1111 @@
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    ...    print(X_train, X_test, y_train, y_test)
+    TRAIN: [1] TEST: [0]
+    [[3 4]] [[1 2]] [2] [1]
+    TRAIN: [0] TEST: [1]
+    [[1 2]] [[3 4]] [1] [2]
+
+    See also
+    --------
+    LeaveOneLabelOut for splitting the data according to explicit,
+    domain-specific stratification of the dataset.
+    """
+
+    def _iter_test_indices(self, X, y=None, labels=None):
+        return range(_num_samples(X))
+
+    def n_splits(self, X, y=None, labels=None):
+        """Returns the number of splitting iterations in the CV iterator."""
+        return _num_samples(X)
+
+
+class LeavePOut(_PartitionIterator):
+    """Leave-P-Out cross validation iterator
+
+    Provides train/test indices to split data in train test sets. This results
+    in testing on all distinct samples of size p, while the remaining n - p
+    samples form the training set in each iteration.
+
+    Note: ``LeavePOut(p)`` is NOT equivalent to ``KFold(n_folds=n // p)``
+    (``n`` = number of samples) which creates non-overlapping
+    test sets.
+
+    Due to the high number of iterations which grows combinatorically with the
+    number of samples this cross validation method can be very costly. For
+    large datasets one should favor KFold, StratifiedKFold or ShuffleSplit.
+
+    Read more in the :ref:`User Guide <split>`.
+
+    Parameters
+    ----------
+    p : int
+        Size of the test sets.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import LeavePOut
+    >>> X = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
+    >>> y = np.array([1, 2, 3, 4])
+    >>> lpo = LeavePOut(2)
+    >>> lpo.n_splits(X)
+    6
+    >>> print(lpo)
+    LeavePOut(p=2)
+    >>> for train_index, test_index in lpo.split(X):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [2 3] TEST: [0 1]
+    TRAIN: [1 3] TEST: [0 2]
+    TRAIN: [1 2] TEST: [0 3]
+    TRAIN: [0 3] TEST: [1 2]
+    TRAIN: [0 2] TEST: [1 3]
+    TRAIN: [0 1] TEST: [2 3]
+    """
+
+    def __init__(self, p):
+        self.p = p
+
+    def _iter_test_indices(self, X, y=None, labels=None):
+        for combination in combinations(range(_num_samples(X)), self.p):
+            yield np.array(combination)
+
+    def n_splits(self, X, y=None, labels=None):
+        """Returns the number of splitting iterations in the CV iterator."""
+        return int(comb(_num_samples(X), self.p, exact=True))
+
+
+class _BaseKFold(with_metaclass(ABCMeta, _PartitionIterator)):
+    """Base class to validate KFold approaches"""
+
+    @abstractmethod
+    def __init__(self, n_folds, shuffle, random_state):
+        self.n_folds = n_folds = int(n_folds)
+
```

Comment:
Thanks to github I have no idea what your last two comments were about lol

On 08/03/2015 01:42 PM, Vlad Niculae wrote:
>
> In sklearn/model_selection/split.py 
> <https://github.com/scikit-learn/scikit-learn/pull/4294#discussion_r36109909>:
>
> > +    def _iter_test_indices(self, X, y=None, labels=None):
> > +        for combination in combinations(range(_num_samples(X)), self.p):
> > +            yield np.array(combination)
> > +
> > +    def n_splits(self, X, y=None, labels=None):
> > +        """Returns the number of splitting iterations in the CV iterator."""
> > +        return int(comb(_num_samples(X), self.p, exact=True))
> > +
> > +
> > +class _BaseKFold(with_metaclass(ABCMeta, _PartitionIterator)):
> > +    """Base class to validate KFold approaches"""
> > +
> > +    @abstractmethod
> > +    def __init__(self, n_folds, shuffle, random_state):
> > +        self.n_folds = n_folds = int(n_folds)
> > +
> Yeah, you’re right.
> … <#>
> On Jul 29, 2015, at 3:20 PM, Andreas Mueller 
> <notifications@github.com> wrote: In sklearn/model_selection/split.py: 
> > + def _iter_test_indices(self, X, y=None, labels=None): > + for 
> combination in combinations(range(_num_samples(X)), self.p): > + yield 
> np.array(combination) > + > + def n_splits(self, X, y=None, 
> labels=None): > + """Returns the number of splitting iterations in the 
> CV iterator.""" > + return int(comb(_num_samples(X), self.p, 
> exact=True)) > + > + > +class _BaseKFold(with_metaclass(ABCMeta, 
> _PartitionIterator)): > + """Base class to validate KFold 
> approaches""" > + > + @abstractmethod > + def __init__(self, n_folds, 
> shuffle, random_state): > + self.n_folds = n_folds = int(n_folds) > + 
> On 06/30/2015 07:16 PM, Vlad Niculae wrote: Since this is a new class, 
> could we move all validation out of |__init__| and into |split|, and 
> change the validation to disallow |n_folds=3.0|? This is not an 
> estimator, so there is not really any reason for that, right? … — 
> Reply to this email directly or view it on GitHub.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/scikit-learn/scikit-learn/pull/4294/files#r36109909>.
>